xayn-ai/src/embedding/utils.rs

```diff
@@ -56,23 +56,26 @@ where
 /// # Panics
 /// Panics if the vectors don't consist solely of real values or their shapes don't match.
 #[allow(dead_code)]
-pub fn pairwise_cosine_similarity<S>(a: &[ArrayBase<S, Ix1>]) -> Array2<f32>
+pub fn pairwise_cosine_similarity<'a, I, S>(iter: I) -> Array2<f32>
 where
-    S: Data<Elem = f32>,
+    I: IntoIterator<Item = &'a ArrayBase<S, Ix1>>,
+    I::IntoIter: Clone,
+    S: Data<Elem = f32> + 'a,
```

Comment:
Do we need the additional complexity and could we not use `ExactSizedIterator`?

Both cloning and iterator and counting it a things prone to accidentally introducing a bunch of overhead.

Reply:
i encountered a situation in the key phrase impl where it's nicer to have it accept an iterator to avoid collecting too much. the `ExactSizedIterator` is a good idea, i'll change that. cloning an iterator is usually quite cheap, for example the iterator for slices is just to pointers big.